docs/network-layer.md

```diff
@@ -0,0 +1,61 @@
+
+It is also possible for a node to be *pinned* to certain other peers. This is a task that is carried out by the node operator when configuring the node. The node will always try to keep a connection the the *pinned* peers.
+
+The node operator can via the RPC interface make the node *join* or *leave* a given *network* (specified by the `NetworkId`).
+
+## Connection Establishment
+When a node and a peer tries to establish a connection, then the node and the other peer will perform a 2-part handshake.
+
+The purpose of the *first part* of the handshake is to create a *secure channel* between the two peers. In practice a [*Noise XX*](https://noiseexplorer.com/patterns/XX/) handshake is being carried out by the two peers. This ensures *authenticity* and *confidentiality* for the rest of the communication.
+
+After completing the first part of the handshake, then the two peers will carry out the second part of the handshake. By checking various meta data, the two peers confirm that they are in fact compatible. Specfic the peers asserts that:
+
+1. The client versions are compatible
+2. The wire versions are compatible
+3. The peers have the same genesis block
+
+When the *handshakes* has been successfully completed then the peers will *promote* each other into their corresponding lists of peers. I.e. the peers are now connected and can communicate with each other, hence the peers can now relay messages between them, carry out *consensus* and *catchup* etc.
+
+## Control Flow
+When the node is running, it performs several housekeeping tasks wrt. networking related matters. More specifically the node checks that its current connections are *healthy* i.e., verifying reasonable latency, removing connections of which a handshake is missing etc. If any connections deem *unhealthy* then the node will close the connection and remove the peer from its set of peers. Note that depending on the level of *unhealthy* the node will perhaps *soft ban* a peer (more on this below).
+
+If the node has exceeded its maximum amount of peers it will remove a random subset of its peers. Note that the node will never remove any of the *pinned peers*. 
+The node then proceeds to check if it should unban any of the *soft banned* peers and does so if that is the case. 
+
+Afterwards the node may try to *re-bootstrap* again (if the node did not recently *bootstrap*). Finally if the node still lacks peers then it will send out *GetPeers* messages to its peers.
+
+The node listens on its established connections and processes the received messages accordingly. If the message was *re-broadcastable*, then the node will relay the message to its own peers such that the message will propagate throughout the network. Beforehand processing the message the node asserts that it has not yet seen the message before or otherwise the node will not process the *duplicated* message. 
```

Comment:
```suggestion
The node listens on its established connections and processes the received messages accordingly. If the message was *re-broadcastable*, then the node will relay the message to its own peers such that the message will propagate throughout the network. Before processing the message the node asserts that it has not yet seen the message before or otherwise the node will not process the *duplicated* message. 
```

Reply:
Perhaps a paragraph about how deduplication works would be useful here.

Reply:
Will elaborate on this.